yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnTimelineUtils.scala

```diff
@@ -0,0 +1,720 @@
+
+/**
+ * Utility methods for timeline classes.
+ */
+private[spark] object YarnTimelineUtils extends Logging {
+
+  /**
+   * What attempt ID to use as the attempt ID field (not the entity ID) when
+   * there is no attempt info
+   */
+  val SINGLE_ATTEMPT = "1"
+
+
+  /**
+   * Exception text when there is no event info data to unmarshall
+   */
+  val E_NO_EVENTINFO = "No 'eventinfo' entry"
+
+  /**
+   * Exception text when there is event info entry in the timeline event, but it is empty
+   */
+
+  val E_EMPTY_EVENTINFO = "Empty 'eventinfo' entry"
+
+  /**
+   * counter incremented on every spark event to timeline event creation,
+   * so guaranteeing uniqueness of event IDs across a single application attempt
+   * (which is implicitly, one per JVM)
+   */
+  val uid = new AtomicLong(System.currentTimeMillis())
+
+  /**
+   * Converts a Java object to its equivalent json4s representation.
+   */
+  def toJValue(obj: Object): JValue = {
+    obj match {
+      case str: String => JString(str)
+      case dbl: java.lang.Double => JDouble(dbl)
+      case dec: java.math.BigDecimal => JDecimal(dec)
+      case int: java.lang.Integer => JInt(BigInt(int))
+      case long: java.lang.Long => JInt(BigInt(long))
+      case bool: java.lang.Boolean => JBool(bool)
+      case map: JMap[_, _] =>
+        val jmap = map.asInstanceOf[JMap[String, Object]]
+        JObject(jmap.entrySet().map { e => (e.getKey() -> toJValue(e.getValue())) }.toList)
+      case array: JCollection[_] =>
+        JArray(array.asInstanceOf[JCollection[Object]].map(o => toJValue(o)).toList)
+      case null => JNothing
+    }
+  }
+
+  /**
+   * Converts a JValue into its Java equivalent.
+   */
+  def toJavaObject(v: JValue): Object = {
+    v match {
+      case JNothing => null
+      case JNull => null
+      case JString(s) => s
+      case JDouble(num) => java.lang.Double.valueOf(num)
+      case JDecimal(num) => num.bigDecimal
+      case JInt(num) => java.lang.Long.valueOf(num.longValue)
+      case JBool(value) => java.lang.Boolean.valueOf(value)
+      case obj: JObject => toJavaMap(obj)
+      case JArray(vals) => {
+        val list = new JArrayList[Object]()
+        vals.foreach(x => list.add(toJavaObject(x)))
+        list
+      }
+    }
+  }
+
+  /**
+   * Converts a json4s list of fields into a Java Map suitable for serialization by Jackson,
+   * which is used by the ATS client library.
+   */
+  def toJavaMap(sourceObj: JObject): JHashMap[String, Object] = {
+    val map = new JHashMap[String, Object]()
+    sourceObj.obj.foreach(f => map.put(f._1, toJavaObject(f._2)))
+    map
+  }
+
+  /**
+   * Convert a timeline event to a spark one. Includes
+   * some basic checks for validity of event payload.
+   * @param event timeline event
+   * @return an unmarshalled event
+   */
+  def toSparkEvent(event: TimelineEvent): SparkListenerEvent = {
+    val info = event.getEventInfo()
+    if (info == null) {
+      throw new IOException(E_NO_EVENTINFO)
```

Comment:
It's just that this doesn't seem like a deser issue; you seem to be asserting that the data retrieved from the ATS matches your expectations. If it were caused by bad serialized data, wouldn't an `IOException` or some deserializer-specific exception be thrown somewhere up the call stack?